Generate weekly coverage report for C++ code

daljit46 opened the pull request:
As mentioned in #2773, this adds a Github Actions workflow to generate a coverage report for C++. It outputs a coverage report after running unit and binary tests. Also generates an html file to view coverage in a browser (using `gcovr`).

daljit46 committed:
Add coverage workflow

daljit46 committed:
REMOVE BEFORE MERGE: trigger on pull request

jdtournier commented:
I suspect this needs to be merged to `dev`...?

daljit46 commented:
This needs to be on `master` for the same reasons as #2763.

Lestropie committed:
Merge branch 'master' into coverage

Lestropie committed:
Various tweaks to C++ coverage testing

Update upload-artifact action to version 4
Update CMake version to 3.22.1 in coverage report generation
Update sccache-action version in coverage.yml
Attempt to use same version as is currently in use on dev branch for CI checks to see if it resolves CI failure:
sccache: error: Server startup failed: cache storage failed to read: Unexpected (permanent) at read => <h2>Our services aren't available right now</h2><p>We're working to restore all services as soon as possible. Please check back soon.</p>
Fix coverage testing build and test invocation
Print software versions in C++ coverage testing workflow
Fix reporting of gcovr version

Lestropie commented:
Attempting to bring this PR up to date and merge. There's a few PRs happening in the "code quality" space where I'm relying a decent amount on testing to ensure that nothing is being unintentionally broken, but testing coverage is not fantastic at the moment. Hoping that with a coverage report I can expedite constructing tests (with existing test data where possible, with new data where necessary) that will interrogate those code paths.

Lestropie committed:
Install libpng-dev in coverage workflow

Lestropie commented:
- [x] Remove line activating coverage testing on pull requests before merging.

Lestropie committed:
Ignore negative hit counts in coverage testing

Attempt to address gcov returning error code due to the following:
gcovr.formats.gcov.parser.NegativeHits: Got negative hit value in gcov line ... caused by a bug in gcov tool, see https://gcc.gnu.org/bugzilla/show_bug.cgi?id=68080. Use option --gcov-ignore-parse-errors with a value of negative_hits.warn, or negative_hits.warn_once_per_file.

Lestropie committed:
Remove pull_request trigger from coverage workflow

Lestropie committed:
Tweaks to C++ coverage analysis workflow

- Rename to be more consistent with Python coverage analysis.
- Use output paths consistent with the Python coverage analysis.
- Activate finite retention time.

daljit46 commented on .github/workflows/coverage_cpp.yml:
```diff
@@ -0,0 +1,84 @@
+      with:
+        name: cpp-coverage-report
+        path: coverage_reports/
+        retrntion-days: 90
```
I think this is a typo.

daljit46 commented:
This looks good to me. @Lestropie It would be nice to have a separate repo for checking out the coverage stats using Github Pages as demonstrated [here](https://github.com/MRtrix3/mrtrix3/issues/2773#issuecomment-1893877613), rather than downloading HTML files (and I believe this should be fairly straightforward to do using Github's APIs).
